app/views/unattended/provisioning_templates/registration/global_registration.erb

```diff
@@ -86,8 +86,23 @@ EOF
 elif [ -f /etc/debian_version ]; then
   <%= save_to_file('/etc/apt/sources.list.d/foreman_registration.list', @repo) %>
 <% if @repo_gpg_key_url.present? -%>
+<%
+  <% # Ubuntu 16.04 does not support GPG public keys in "/etc/apt/trusted.gpg.d/" -%>
+  use_apt_key =
+    if @host.operatingsystem.name == 'Ubuntu' && @host.operatingsystem.major.split('.')[0].to_i == 16
+      true
+    else
+      false
+    end
+%>
+
   apt-get -y install ca-certificates gpg
+  <% if use_apt_key %>
+  curl --silent --show-error <%= shell_escape @repo_gpg_key_url %> -O - | apt-key add -
+  <% else %>
   curl --silent --show-error --output /etc/apt/trusted.gpg.d/client.asc <%= shell_escape @repo_gpg_key_url %>
```

Comment:
Is this really `.asc`? If so, Ubuntu 16.04 should support this. Otherwise the file should be named `client.gpg`. And perhaps also qualify name it `foreman_registration.$extension`? Or even derive it from `repo_gpg_key_url`? Something like `File.basename(@repo_gpg_key_url)` probably works.